Blog: Policy as Code for Any Cloud Providers (#4416)
pulumi/docs: content/blog/multicloud-policy-as-code/index.md

---
title: "Policy as Code for Any Cloud Provider"
date: 2020-10-22T11:43:21-05:00

> meta_desc: "Create policies for all clouds from AWS, Azure, Google Cloud Platform, and Kubernetes"

Reviewer: ```suggestion
meta_desc: "Create policies for all clouds from AWS, Azure, Google Cloud, and Kubernetes"
```

meta_image: meta.png
authors:
    - sophia-parafina
tags:
    - policy as code
---

Policies protect your infrastructure by controlling access, set limits that reduce the blast radius of an incident, and manage infrastructure operations. Policies are commonly created through a form on a cloud provider's administrative console, making replicating or versioning the policy more difficult. With Policy as Code, you can apply software engineering practices such as automated testing, deployment, and version control when creating policies.

> CrossGuard is Pulumi's Policy as Code solution that lets you create, verify, apply, and enforce policies. Policies are standalone packages that can be run against any Pulumi stack. That means your policies are language agnostic and work with any language supported by Pulumi. Policy Packages are policy bundles that evaluate every resource in your stack, whether deployed in AWX, Azure, Google Cloud Platform, or Kubernetes.

Reviewer: Nit: From my experience, these days Google appears to largely call it just "Google Cloud" rather than "Google Cloud Platform"...

```suggestion
CrossGuard is Pulumi's Policy as Code solution that lets you create, verify, apply, and enforce policies. Policies are standalone packages that can be run against any Pulumi stack. That means your policies are language agnostic and work with any language supported by Pulumi. Policy Packages are policy bundles that evaluate every resource in your stack, whether deployed in AWS, Azure, Google Cloud, or Kubernetes.
```

<!--more-->

The key features of CrossGuard are:

> - Policy SDK for coding custom policies using TypeScript/Javascript or Python

Reviewer: ```suggestion
- Policy SDK for coding custom policies using TypeScript/JavaScript or Python
```

Author: yes, I'll update meta.png

    EnforcementLevel,
    PolicyPack,
    ReportViolation,
    ResourceValidationArgs,
    ResourceValidationPolicy,
)

def storage_container_no_public_read_validator(args: ResourceValidationArgs, report_violation: ReportViolation):
    if args.resource_type == "azure:storage/container:Container" and "containerAccessType" in args.props:
        access_type = args.props["containerAccessType"]
        if access_type == "blob" or access_type == "container":
            report_violation(
                "Azure Storage Container must not have blob or container access set. " +
                "Read more about read access here: " +
                "https://docs.microsoft.com/en-us/azure/storage/blobs/storage-manage-access-to-resources")

storage_container_no_public_read = ResourceValidationPolicy(
    name="storage-container-no-public-read",
    description="Prohibits setting the public permission on Azure Storage Blob Containers.",
    validate=storage_container_no_public_read_validator,
)

PolicyPack(
    name="azure-python",
    enforcement_level=EnforcementLevel.MANDATORY,
    policies=[
        storage_container_no_public_read,
    ],
)
```

{{% /choosable %}}
{{< /chooser >}}

> ### Google Cloud Platform

Reviewer: ```suggestion
### Google Cloud
```

    enforcement_level=EnforcementLevel.MANDATORY,
    policies=[
        no_public_services,
    ],
)
```

{{% /choosable %}}
{{< /chooser >}}

## All Clouds and all resources

Policy as Code is a powerful tool for protecting and managing your infrastructure. As you can see, Pulumi supports policies for all resources on all clouds. We have examples and articles to get you started with Policy as Code.

- [Policy Pack Examples](https://github.com/pulumi/examples/tree/master/policy-packs)
- [Get Started with Policy as Code]({{< relref "/docs/get-started/crossguard" >}})
- [New Policy as Code Capabilities with CrossGuard]({{< relref "/blog/crossguard-2-0" >}})
- [Automatically Enforcing AWS Resource Tagging Policies]({{< relref "/blog/automatically-enforcing-aws-resource-tagging-policies" >}})
- [Manage Any Infrastructure with Policy as Code]({{< relref "/blog/automatically-enforcing-aws-resource-tagging-policies" >}})
- [Enforcing Different Kinds of Policies for Cloud Resources]({{< relref "/blog/enforcing-different-kinds-of-policies-for-cloud-resources" >}})
- [Running AWS IAM Access Analyzer at Deployment Time]({{< relref "/blog/aws-iam-access-analyzer-and-crossguard" >}})

> [Authoring CrossGuard Policy with Open Policy Agent (OPA)]({{< relref "/blog/opa-support-for-crossguard" >}})

Reviewer: ```suggestion
- [Authoring CrossGuard Policy with Open Policy Agent (OPA)]({{< relref "/blog/opa-support-for-crossguard" >}})
```
